scripts/reactor_deployment.py

```diff
@@ -0,0 +1,496 @@
+                    if next_difference <= -(ar_dict[reactor][0] + 1):
+                        continue
+                    else:
+                        # Check if a cap exists.
+                        if type(ar_dict[reactor][3]) == list:
+                            # Check if the cap will be exceeded by adding
+                            # another reactor.
+                            if ar_dict[reactor][3][year] >= \
+                              (df.loc[year, f'num_{reactor}'] + 1):
+                                df.loc[year, f'num_{reactor}'] += 1
+                                # Update remaining capacity to be met.
+                                cap_difference -= ar_dict[reactor][0]
+                        # If there isn't a cap add another reactor.
+                        else:
+                            df.loc[year, f'num_{reactor}'] += 1
+                            # Update remaining capacity to be met.
+                            cap_difference -= ar_dict[reactor][0]
+
+    # account for decommissioning with a direct replacement
+    df = direct_decom(df, ar_dict)
+
+    # Now calculate the total capacity each year (includes capacity from a
+    # replacement reactor that is new that year, but not new overall because it
+    # is replacing itself).
+    df = num_react_to_cap(df, ar_dict)
+
+    return df
+
+
+def rand_deployment(df, base_col, ar_dict,
+                    set_seed=False, rough=True, tolerance=5):
+    """
+    This function randomly deploys reactors from the dictionary of
+    reactors to meet a capacity need.
+
+    There are two implementations:
+    1) rough (rough=True): if a reactor greater than the remaining
+    capacity is proposed, the deployment ends.
+    2) [IN-PROGRESS] complete (rough=False): the algorithm keeps
+    trying to deploy randomly selected reactors until the capacity demand
+    has been met.
+
+    Parameters
+    ----------
+    df: pandas dataframe
+        The dataframe of capacity information.
+    base_col: str
+        The string name corresponding to the column of capacity that the
+        algorithm is deploying reactors to meet.
+    ar_dict: dictionary
+        A dictionary of reactors with information of the form:
+        {reactor: [Power (MWe), capacity_factor (%), lifetime (yr)]}.
+    set_seed: bool
+        A True/False value that determines whether the seed used for the random
+        number is set or varies based on time.
+    rough: bool
+        A True/False value that determines whether the initial deployment is
+        rough or complete.
+    tolerance: float/int
+        The capacity tolerance to which reactors are deployed in the complete
+        deployment case (i.e., when rough=False). Without this, convergence
+        is tricky to achieve.
+    """
+
+    # initialize the number of reactor columns
+    for reactor in ar_dict.keys():
+        if f'num_{reactor}' not in df:
+            df[f'num_{reactor}'] = 0
+        else:
+            pass
```

Comment:
If you make this into a class method, this could be a function call rather than repeated